### PR TITLE
Have RegExp constructor process RegExp-like objects correctly

### LINKED ISSUE #577: RegExp constructor should process RegExp-like Objects correctly
We differ from the spec in two aspects.

First, when the constructor is called as a function, we should use the `constructor` property of the passed RegExp-like object to determine whether we should return the argument as is. Kangax test case:

```JS
var re = /./;
re[Symbol.match] = false;
var foo = {constructor: RegExp};
foo[Symbol.match] = true;
return RegExp(re) !== re && RegExp(foo) === foo;
```

Second, we should use the `source` and `flags` properties of the passed RegExp-like object to initialize the new RegExp object. Kangax test case:

```JS
// RegExp -> Get -> [[Get]]
var get = [];
var re = { constructor: null };
re[Symbol.match] = true;
var p = new Proxy(re, { get: function(o, k) { get.push(k); return o[k]; }});
RegExp(p);
return get[0] === Symbol.match && get.slice(1) + '' === "constructor,source,flags";
```

### PR DESCRIPTION
**Pull request by goyakin**
Fixes #577

### CODE REVIEW

**Comment by goyakin** [comment]
@akroshg & @abchatra, could you take a look?

File: lib/Runtime/Library/JavascriptRegularExpression.cpp [hunk 1]
```diff
@@ -258,48 +258,61 @@ namespace Js
         {
             pattern = scriptContext->GetLibrary()->GetEmptyRegexPattern();
         }
-        else if (JavascriptRegExp::Is(args[1]))
+        else if (JavascriptRegExp::IsRegExpLike(args[1], scriptContext))
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptRegularExpression.cpp]: will you run the benchmark to verify the perf? (IsRegExpLike seems to costly at this point).

**Comment by akroshg** [comment]
code wise looks good to me.

File: lib/Runtime/Library/JavascriptRegularExpression.cpp (hunk 1 above)
Comment (goyakin) [inline lib/Runtime/Library/JavascriptRegularExpression.cpp]: Benchmarks don't show any difference in perf.

**Pull request merged**